### PR TITLE
Uniform buffer objects

### PR DESCRIPTION
**Pull request by alexcristici**
Uniform Buffer Object Implementation.

### CODE REVIEW

**Review by TimSylvester — APPROVED** [review]

**Review by ovivoda — APPROVED** [review]

**Review by mwilsnd — APPROVED** [review]

File: include/mbgl/gfx/uniform_block.hpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+    /// Get map of elements.
+    const UniformBlockMap& getMap() const { return uniformBlockMap; }
+
+    /// Number of elements
+    std::size_t size() const { return uniformBlockMap.size(); }
+
+    /// Get an uniform block element.
+    /// Returns a pointer to the element on success, or null if the uniform block doesn't exists.
+    UniformBlock* get(const std::string& name) const;
```
Comment (sjg-wdw) [inline include/mbgl/gfx/uniform_block.hpp:63]: Raw pointers don't bother me all that much, but I can see they bother some.
Should the UniformBlock* be some other C++ structure on return?

Comment (TimSylvester) [reply include/mbgl/gfx/uniform_block.hpp:63]: I think this is based on the attribute stuff I did.  It's a `unique_ptr` internally and we don't return ownership, only a temporary reference, but it's optional so it's a pointer.  The nicer semantics would be a `shared_ptr`, but there's no actual need for the refcounting.  Maybe an `optional<reference_wrapper<>>`?

Comment (mwilsnd) [reply include/mbgl/gfx/uniform_block.hpp:63]: We could return a `const std::unique_ptr<>&` temporary which would better convey ownership.

**Comment by sjg-wdw** [comment]
This looks good.  I think we can build on this.

**Review by sjg-wdw — APPROVED** [review]

File: include/mbgl/gfx/uniform_block.hpp (hunk 1 above)
Comment (sjg-wdw) [reply include/mbgl/gfx/uniform_block.hpp:63]: That feels more C++

Comment (alexcristici) [reply include/mbgl/gfx/uniform_block.hpp:63]: I need to return `nullptr`, not sure if this is possible for this return `const std::unique_ptr<>&`.

Comment (TimSylvester) [reply include/mbgl/gfx/uniform_block.hpp:63]: works for me

Comment (mwilsnd) [reply include/mbgl/gfx/uniform_block.hpp:63]: A more C++-y way, but you could return an optional holding a reference wrapper. Ex: https://godbolt.org/z/3xdrnE3dG

Comment (alexcristici) [reply include/mbgl/gfx/uniform_block.hpp:63]: Yes, maybe we should create a global one somewhere.

Comment (alexcristici) [reply include/mbgl/gfx/uniform_block.hpp:63]: So we should used `optional<reference_wrapper<>>` as also Tim mentioned?

**Review by louwers — APPROVED** [review]

File: include/mbgl/gfx/uniform_block.hpp (hunk 1 above)
Comment (mwilsnd) [reply include/mbgl/gfx/uniform_block.hpp:63]: I'm fine with the pointer for now. The reference wrapper is more verbose but does help convey the returned reference is temporary.

Comment (stefankarschti) [reply include/mbgl/gfx/uniform_block.hpp:63]: I'm in favor of `const std::unique_ptr<>&`

File: include/mbgl/gfx/uniform_block.hpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
+#pragma once
+
+#include <string>
```
Comment (mwilsnd) [inline include/mbgl/gfx/uniform_block.hpp:3]: This file doesn't include everything it needs to build the test runner on linux. Include unordered_map and memory here.

File: include/mbgl/gfx/uniform_buffer.hpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <string>
```
Comment (mwilsnd) [inline include/mbgl/gfx/uniform_buffer.hpp:3]: ditto

File: src/mbgl/gl/drawable_gl.cpp [hunk 4]
```diff
@@ -67,12 +71,42 @@ void DrawableGL::setVertexArray(gl::VertexArray&& vertexArray_,
...
+
 void DrawableGL::resetColor(const Color& newColor) {
     if (auto* colorAttr = impl->vertexAttributes.get("a_color")) {
         colorAttr->clear();
         colorAttr->set(0, colorAttrValue(newColor));
     }
 }
 
+void DrawableGL::bindUniformBuffers() const {
+    if (const auto& shader = getShader()) {
```
Comment (mwilsnd) [inline src/mbgl/gl/drawable_gl.cpp:90]: `shader` shadows member variable here (-Werror=shadow).

File: src/mbgl/gl/drawable_gl.cpp [hunk 5]
```diff
@@ -67,12 +71,42 @@ void DrawableGL::setVertexArray(gl::VertexArray&& vertexArray_,
...
+            if (!uniformBuffer) {
+                continue;
+            }
+            element.second->bindBuffer(*uniformBuffer);
+        }
+    }
+}
+
+void DrawableGL::unbindUniformBuffers() const {
+    if (const auto& shader = getShader()) {
```
Comment (mwilsnd) [inline src/mbgl/gl/drawable_gl.cpp:103]: ditto

File: src/mbgl/gl/uniform_block_gl.cpp [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+#include <mbgl/gl/defines.hpp>
+#include <mbgl/platform/gl_functions.hpp>
+
+namespace mbgl {
+namespace gl {
+
+using namespace platform;
+
+void UniformBlockGL::bindBuffer(const gfx::UniformBuffer& uniformBuffer) {
+    assert(size == uniformBuffer.getSize());
```
Comment (mwilsnd) [inline src/mbgl/gl/uniform_block_gl.cpp:13]: `#include <cassert>`

**Review by mwilsnd — CHANGES_REQUESTED** [review]

**Review by mwilsnd — CHANGES_REQUESTED** [review]

File: src/mbgl/renderer/renderer_impl.cpp [hunk 7]
```diff
@@ -235,8 +235,10 @@ void Renderer::Impl::render(const RenderTree& renderTree) {
             }
 
             if (auto& shader = drawable.getShader()) {
```
[the remark this answers is not in the document]
Comment (alexcristici) [reply src/mbgl/renderer/renderer_impl.cpp:237]: I removed the `if`, not used.

File: include/mbgl/gfx/uniform_block.hpp (hunk 1 above)
Comment (alexcristici) [reply include/mbgl/gfx/uniform_block.hpp:65]: I changed it to `const std::unique_ptr<>&`.

**Review by stefankarschti — APPROVED** [review]

**Review by mwilsnd — APPROVED** [review]

**Pull request merged**